pydrofoil/real.py

```diff
@@ -0,0 +1,99 @@
+import sys
+from rpython.rlib.rbigint import rbigint, _divrem as bigint_divrem
+from rpython.rlib.rarithmetic import r_uint, intmask, string_to_int, ovfcheck, \
+        int_c_div, int_c_mod, r_ulonglong
+
+class Real(object):
+    def __init__(self, num, den):
+        self.num = num
+        self.den = den
+
+    @staticmethod
+    def fromint(num, den=1):
+        sign = num/abs(num)*den/abs(den)
+        num = rbigint.fromint(abs(num))
+        den = rbigint.fromint(abs(den)*sign)
+        # assert den.int_ne(0)
+        assert not den.int_eq(0), "denominator cannot be 0"
+        return Real(num.div(num.gcd(den)), den.div(num.gcd(den)))
+    
+    def add(self, other):
+        num_new_1 = self.num.mul(other.den)
+        num_new_2 = other.num.mul(self.den)
+        self.den = self.den.mul(other.den)
+        self.num = num_new_1.add(num_new_2)
+        sign = self.num.div(self.num.abs()).mul(self.den.div(self.den.abs()))
+        self.num = self.num.abs()
+        self.den = self.den.abs().mul(sign)
```

Comment:
I am not convinced by all these `abs` calls. If you switch to the denominator being always positive, you can remove the `den.abs` calls immediately. I am also not convinced that you need to compute the sign explicitly, it should work out correctly automatically if the rest of the code is correct.

also, you are mutating `self` here as well.

Reply:
yes I would check this, I use these abs calls just to make sure the negative sign always goes to denominator, but it would be easier for another way around, I would change it.